### PR TITLE
Filter out passwords before sending params to bugsnag

### PR DESCRIPTION
**Pull request by bitboxer**
This adds a password filter.

Disclaimer: I hope this is the correct way to do this. I searched online for a way to recursively change a map and found code that did what I wanted in an article 😀. Please point out all the things that could be better 👍.

### CODE REVIEW

File: lib/bugsnex/plug.ex [hunk 1]
```diff
@@ -75,4 +77,21 @@ defmodule Bugsnex.Plug do
     {:ok, hostname} = :inet.gethostname
     to_string(hostname)
   end
+
+  defp filter_parameters(params) do
+    do_filter(params, Application.get_env(:bugsnex, :filter_params, @default_filter_params))
+  end
+
+  defp do_filter(%{__struct__: mod} = struct, _params) when is_atom(mod), do: struct
```
Comment (PragTob) [inline lib/bugsnex/plug.ex]: also `_params` here is rather `_filter_params` (params is the first argument)

File: lib/bugsnex/plug.ex [hunk 2]
```diff
@@ -75,4 +77,21 @@ defmodule Bugsnex.Plug do
...
+  defp do_filter(%{} = map, params) do
```
Comment (PragTob) [inline lib/bugsnex/plug.ex]: same thing, first argument is actual params, second is `filter_params` or well, maybe a better name is `params_to_filter` or `param_keys_to_filter` or something like that, `params` makes me think it's the actual parameters  (which gets confusing when we check for membership)

File: lib/bugsnex/plug.ex [hunk 3]
```diff
@@ -75,4 +77,21 @@ defmodule Bugsnex.Plug do
...
+    Enum.into map, %{}, fn {k, v} ->
+      if is_binary(k) && Enum.member?(params, k) do
```
Comment (PragTob) [inline lib/bugsnex/plug.ex]: I don't think we need to check that they key is binary, we can just check with any key if it's a member of our forbidden/to be filtered parameter keys, worst case it returns false :)

File: lib/bugsnex/plug.ex [hunk 4]
```diff
@@ -75,4 +77,21 @@ defmodule Bugsnex.Plug do
...
     to_string(hostname)
   end
+
+  defp filter_parameters(params) do
+    do_filter(params, Application.get_env(:bugsnex, :filter_params, @default_filter_params))
+  end
+
+  defp do_filter(%{__struct__: mod} = struct, _params) when is_atom(mod), do: struct
+  defp do_filter(%{} = map, params) do
+    Enum.into map, %{}, fn {k, v} ->
```
Comment (PragTob) [inline lib/bugsnex/plug.ex]: Small thing you don't have to do, but I really like being more verbose aka `key`/`value` here :) Helps with readability a bit further down imo

File: lib/bugsnex/plug.ex [hunk 5]
```diff
@@ -75,4 +77,21 @@ defmodule Bugsnex.Plug do
...
+      if is_binary(k) && Enum.member?(params, k) do
+        {k, "[FILTERED]"}
+      else
+        {k, do_filter(v, params)}
+      end
+    end
+  end
+  defp do_filter([_|_] = list, params), do: Enum.map(list, &do_filter(&1, params))
```
Comment (PragTob) [inline lib/bugsnex/plug.ex]: nice one, I probably would have forgotten about the list case! :+1:

File: lib/bugsnex/plug.ex (hunk 1 above)
Comment (PragTob) [inline lib/bugsnex/plug.ex]: This avoids altering structs, which means that if somehow a struct ends up in the params (which it generally shouldn't) we wouldn't sanitize its keys. But it's keys would be atoms. And it should never happen. We are probably fine. Just thinking out loud.

File: test/bugsnex/plug_test.exs [hunk 6]
```diff
@@ -60,10 +60,10 @@ defmodule Bugsnex.PlugTest do
   end
 
   test "build_plug_env/2" do
-    conn = conn(:get, "/bang?foo=bar")
+    conn = conn(:get, "/bang?foo=bar&password=password&password_confirmation=password")
     plug_env = %{request: Bugsnex.Plug.build_request_data(conn),
                  context: "/bang",
-                 params: %{"foo" => "bar"},
+                 params: %{"foo" => "bar", "password" => "[FILTERED]", "password_confirmation" => "[FILTERED]"},
```
Comment (PragTob) [inline test/bugsnex/plug_test.exs]: the code above does a lot of things that are not tested here (it avoids altering structs, it works for nested structures, it works for list of things...) which I think would all be nice to test here :) If you don't wanna do it, let me know and I/we will add the tests.

**Comment by bitboxer** [comment]
I fixed the names, should have done that already instead of copy and pasting/modifying an example code 😉 . I tried to add deeper maps in the test, but failed to find a good way to do this. Conn and Plug testing is still a bit new to me. It would be awesome if you could guide me a bit there.

**Comment by PragTob** [comment]
Hello,

sorry for the down time. 

You could add another POST end point to the tested plug application, or you could just continue with encoded param string that you can generate and then expect that what arrives has all the desired values filtered.

--> use this in the request URL for continuing with GET requests: 

```
Plug.Conn.Query.encode(%{foo: "bar", map: %{a: 1, b: [%{pw: "secret"}]}})
# => "foo=bar&map[a]=1&map[b][][pw]=secret"
```

that should work :)

**Comment by bitboxer** [comment]
Thanks! Will do this on monday when I am back at working 👍

**Comment by bitboxer** [comment]
Okay, added that map test.

**Comment by PragTob** [comment]
I think I'll add one or more test cases but that's quicker if I do it right now :)

Thanks!

**Pull request merged**